src/main/java/io/cdap/plugin/gcp/bigquery/connector/BigQueryConnector.java

```diff
@@ -233,19 +242,68 @@ private BigQuery getBigQuery(String project) throws IOException {
   }
 
   private List<StructuredRecord> getTableData(BigQuery bigQuery, String datasetProject, String dataset, String table,
-    int limit)
+    int limit, String sampleType, @Nullable String strata, @Nullable Long timeoutMs)
     throws IOException {
-    String query =
-      String.format("SELECT * FROM `%s.%s.%s` LIMIT %d", datasetProject, dataset, table, limit);
+    if (sampleType == null) {
+      sampleType = "first";
+    }
+    String tableName = String.format("`%s.%s.%s`", datasetProject, dataset, table);
+    String query;
+    switch (SampleType.fromString(sampleType)) {
+      case RANDOM:
+        query = String.format("WITH table AS (\n" +
+                                "  SELECT *, RAND() AS r\n" +
+                                "  FROM %s\n" +
+                                "  WHERE RAND() < 2*%d/(SELECT COUNT(*) FROM %s)\n" +
+                                ")\n" +
+                                "SELECT * EXCEPT (r)\n" +
+                                "FROM table\n" +
+                                "ORDER BY r\n" +
+                                "LIMIT %d", tableName, limit, tableName, limit);
+        break;
+      case STRATIFIED:
+        if (strata == null) {
+          throw new IllegalArgumentException("No strata column given.");
+        }
+        query = String.format("SELECT * EXCEPT (`seqnum`, `cnt`)\n" +
+                        "FROM (\n" +
+                        "SELECT *, row_number() OVER (ORDER BY %s, RAND()) AS seqnum,\n" +
+                        "COUNT(*) OVER () as cnt,\n" +
+                        "FROM %s\n" +
+                        ") %s\n" +
+                        "WHERE MOD(seqnum, CAST(cnt / %d AS INT64)) = 1\n" +
+                        "ORDER BY %s\n" +
+                        "LIMIT %d",
+                strata, tableName, tableName, limit, strata, limit);
+        break;
+      default:
+        query =
+          String.format("SELECT * FROM %s LIMIT %d", tableName, limit);
+    }
+
     QueryJobConfiguration queryConfig = QueryJobConfiguration.newBuilder(query).build();
     String id = UUID.randomUUID().toString();
     JobId jobId = JobId.of(id);
     Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(jobId).build());
+
     // Wait for the job to finish
     try {
-      queryJob = queryJob.waitFor();
+      if (timeoutMs == null) {
+        queryJob = queryJob.waitFor();
+      } else {
+        // We use the Future object to set a time limit on the query
+        Job tempQueryJob = queryJob;
+        Future<Job> future = Executors.newCachedThreadPool().submit(() -> tempQueryJob.waitFor());
```

Comment:
You don't need to set up your own executor to do this. The` waitFor()` method in the QueryJob class already has logic to wait for the job for a certain period of time.

See https://javadoc.io/static/com.google.cloud/google-cloud-bigquery/1.137.1/com/google/cloud/bigquery/Job.html#waitFor-com.google.cloud.RetryOption...- for information on how to set the timeout for this method.

Reply:
This method returns either:

`null` if the job has been deleted.
an instance of `Job` containing the latest information available for this job. You can use this to check if the job has has an error (and display this), if the job is still running (in which case you have to cancel it and notify of the timeout) or success (in which case you handle the results)

Reply:
Great point -- I remember seeing this before, but I can't remember why I didn't use it. Maybe I thought it was for network timeouts? Anyway, this simplified the code a lot!